content/blog/2020/10/2020-10-18-a-sustainable-pattern-with-shared-library.adoc

```diff
@@ -0,0 +1,262 @@
+
+== The Problems ==
+
+Large **Jenkinsfiles in every repository containing duplicated code**. It seems common that the `Jenkinsfile`:s in every repository contains much more than just the things that are unique for that repository. The link:/doc/book/pipeline/shared-libraries/[shared libraries] feature may not be used, or it is used but not with an optimal pattern.
+
+**Installation specific Jenkinsfile:s** that only work with one specific Jenkins installation. Sometimes I see multiple `Jenkinsfile`:s, one for each purpose or Jenkins installation. 
+
+**No documentation** and/or no natural place to write documentation.
+
+**Development is slow**. Adding new features to repositories is a time consuming task. I want to be able to push features to 1000+ repositories without having to update their `Jenkinsfile`:s.
+
+**No flexible way of doing feature toggling**. When maintaining a large number of repositories it is sometimes nice to introduce a feature to a subset of those repositories. If that works well, the feature is introduced to all repositories.
+
+== The Solution ==
+
+My solution is a pattern that is inspired by how the link:https://github.com/jenkinsci:[Jenkins organization on GitHub] does it with its link:https://github.com/jenkins-infra/pipeline-library/blob/master/vars/buildPlugin.groovy[buildPlugin()]. But it is not exactly the same.
+
+=== Shared Library ===
+
+Here is how I organize my link:/doc/book/pipeline/shared-libraries/[shared libraries].
+
+==== Jenkinsfile ====
+
+I put this in the `Jenkinsfile`:s:
+
+```groovy
+buildRepo()
+```
+
+==== Default Configuration ====
+
+I provide a default configuration that any repository will get, if no other configuration is given in `buildRepo()`.
+
+I create a `vars/getConfig.groovy` with:
+
+```groovy
+def call(givenConfig = [:]) {
+  def defaultConfig = [
+    /**
+      * The Jenkins node, or label, that will be allocated for this build.
+      */
+    "jenkinsNode": "BUILD",
+    /**
+      * All config specific to NPM repo type.
+      */
+    "npm": [
+      /**
+        * Whether or not to run Cypress tests, if there are any.
+        */
+      "cypress": true
+    ],
+    "maven": [
+      /**
+        * Weather or not to run integration tests, if there are any.
+        */
+      "integTest": true
+    ]
+  ]
+  // https://e.printstacktrace.blog/how-to-merge-two-maps-in-groovy/
+  def effectiveConfig merge(defaultConfig, givenConfig)
+  println "Configuration is documented here: https://whereverYouHos/getConfig.groovy"
+  println "Default config: " + defaultConfig
+  println "Given config: " + givenConfig
+  println "Effective config: " + effectiveConfig
+  return effectiveConfig
+}
+```
+
+==== Build Plan ====
+
+I construct a build plan as early as possible. Taking decisions on what till be done in this build. So that the rest of the code becomes more streamlined.
```

Comment:
```suggestion
I construct a build plan as early as possible. Taking decisions on what will be done in this build. So that the rest of the code becomes more streamlined.
```